packages/logging/src/analytics-helpers.ts

```diff
@@ -119,4 +159,208 @@ export class ToggleableAnalytics implements MongoshAnalytics {
+    // created by long running process (longer than staleDuration) we make sure
+    // that another process doesn't consider lockfile stale
+    intervalId = setInterval(() => {
+      const now = Date.now();
+      fs.promises.utimes(lockfilePath, now, now).catch(() => {});
+    }, staleDuration / 2);
+    // eslint-disable-next-line chai-friendly/no-unused-expressions
+    intervalId.unref?.();
+    return unlock;
+  } catch (e) {
+    if ((e as any).code !== 'EEXIST') {
+      throw e;
+    }
+    const stats = await fs.promises.stat(lockfilePath);
+    // To make sure that the lockfile is not just a leftover from an unclean
+    // process exit, we check whether or not it is stale
+    if (Date.now() - stats.mtimeMs > staleDuration) {
+      await fs.promises.rmdir(lockfilePath);
+      return lockfile(filepath, staleDuration);
+    }
+    throw new Error(`File ${filepath} already locked`);
+  }
+}
+
+export class ThrottledAnalytics implements MongoshAnalytics {
+  private trackQueue = new Queue<AnalyticsTrackMessage>((message) => {
+    if (this.shouldEmitAnalyticsEvent()) {
+      this.target.track(message);
+      this.throttleState.count++;
+    }
+  });
+  private target: ThrottledAnalyticsOptions['target'] = new NoopAnalytics();
+  private currentUserId: string | null = null;
+  private throttleOptions: ThrottledAnalyticsOptions['throttle'] = null;
+  private throttleState = { count: 0, timestamp: Date.now() };
+  private restorePromise: Promise<void> = Promise.resolve();
+  private unlock: () => Promise<void> = () => Promise.resolve();
+
+  constructor({ target, throttle }: Partial<ThrottledAnalyticsOptions> = {}) {
+    this.target = target ?? this.target;
+    this.throttleOptions = throttle ?? this.throttleOptions;
+  }
+
+  get metadataPath() {
+    if (!this.throttleOptions) {
+      throw new Error(
+        'Metadata path is not avaialble if throttling is disabled'
+      );
+    }
+
+    if (!this.currentUserId) {
+      throw new Error('Metadata path is not avaialble if userId is not set');
+    }
+
+    const {
+      throttleOptions: { metadataPath },
+      currentUserId: userId
+    } = this;
+
+    return path.resolve(metadataPath, `am-${userId}.json`);
+  }
+
+  identify(message: AnalyticsIdentifyMessage): void {
+    if (this.currentUserId) {
+      throw new Error('Identify can only be called once per user session');
+    }
+    this.currentUserId = message.userId ?? message.anonymousId;
+    this.restorePromise = this.restoreThrottleState().then((enabled) => {
+      if (!enabled) {
+        this.trackQueue.disable();
+        return;
+      }
+      if (this.shouldEmitAnalyticsEvent()) {
+        this.target.identify(message);
+        this.throttleState.count++;
+      }
+      this.trackQueue.enable();
+    });
+  }
+
+  track(message: AnalyticsTrackMessage): void {
+    this.trackQueue.push(message);
+  }
+
+  private async restoreThrottleState(): Promise<boolean> {
```

Comment:
```suggestion
  // This method must not throw exceptions, since there is nothing to handle them.
  private async restoreThrottleState(): Promise<boolean> {
```

Reply:
Yeah, makes sense to document this better, thanks! (added some more details to the comment)